dds/DCPS/XTypes/DynamicVwrite.cpp

```diff
@@ -0,0 +1,1042 @@
+      write_enum(vw, disc_type, val, treat_disc_as);
+    } else {
+      vw.write_int16(val);
+    }
+    break;
+  }
+  case TK_UINT16: {
+    CORBA::UShort val = 0;
+    rc = value->get_uint16_value(val, id);
+    if (!check_rc(rc, id, disc_tk, "vwrite_discriminator")) {
+      return;
+    }
+    disc_val = static_cast<CORBA::Long>(val);
+    vw.write_uint16(val);
+    break;
+  }
+  case TK_INT32: {
+    rc = value->get_int32_value(disc_val, id);
+    if (!check_rc(rc, id, disc_tk, "vwrite_discriminator")) {
+      return;
+    }
+    if (disc_tk == TK_ENUM) {
+      write_enum(vw, disc_type, disc_val, treat_disc_as);
+    } else {
+      vw.write_int32(disc_val);
+    }
+    break;
+  }
+  case TK_UINT32: {
+    CORBA::ULong val = 0;
+    rc = value->get_uint32_value(val, id);
+    if (!check_rc(rc, id, disc_tk, "vwrite_discriminator")) {
+      return;
+    }
+    disc_val = static_cast<CORBA::Long>(val);
+    vw.write_uint32(val);
+    break;
+  }
+  case TK_INT64: {
+    CORBA::LongLong val = 0;
+    rc = value->get_int64_value(val, id);
+    if (!check_rc(rc, id, disc_tk, "vwrite_discriminator")) {
+      return;
+    }
+    disc_val = static_cast<CORBA::Long>(val);
+    vw.write_int64(val);
+    break;
+  }
+  case TK_UINT64: {
+    CORBA::ULongLong val = 0;
+    rc = value->get_uint64_value(val, id);
+    if (!check_rc(rc, id, disc_tk, "vwrite_discriminator")) {
+      return;
+    }
+    disc_val = static_cast<CORBA::Long>(val);
+    vw.write_uint64(val);
+    break;
+  }
+  default:
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_discriminator: Invalid discriminator type %C\n",
+                 typekind_to_string(disc_tk)));
+    }
+  }
+}
+
+void vwrite_union(ValueWriter& vw, DDS::DynamicData_ptr value, const DDS::DynamicType_var& dt)
+{
+  DDS::TypeDescriptor_var type_desc;
+  DDS::ReturnCode_t rc = dt->get_descriptor(type_desc);
+  if (rc != DDS::RETCODE_OK) {
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_union: get_descriptor failed (%C)\n",
+                 retcode_to_string(rc)));
+    }
+    return;
+  }
+
+  const Extensibility extensibility = XTypes::dds_to_opendds_ext(type_desc->extensibility_kind());
+  vw.begin_union(extensibility);
+
+  // Discriminator
+  DDS::DynamicTypeMember_var dtm;
+  rc = dt->get_member(dtm, XTypes::DISCRIMINATOR_ID);
+  if (rc != DDS::RETCODE_OK) {
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_union: get_member failed (%C)\n",
+                 retcode_to_string(rc)));
+    }
+    return;
+  }
+  DDS::MemberDescriptor_var disc_md;
+  rc = dtm->get_descriptor(disc_md);
+  if (rc != DDS::RETCODE_OK) {
+    if (log_level >= LogLevel::Warning) {
+      ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_union: get_descriptor failed (%C)\n",
+                 retcode_to_string(rc)));
+    }
+    return;
+  }
+  // Discriminator Id is always 0?
```

Comment:
Discriminator doesn't have Id in TypeObject and when converting to DynamicType, we use DISCRIMINATOR_ID as a special Id used to access it. When serializing the union, it needs to get the actual Id to write the Emheader but we don't have that. I think we're using the Id of 0 everywhere for the discriminator for the purpose of serialization right now.

I'm not sure if it's specified in the IDL spec, but it seems to me discriminator is different from the regular members since it's always the first thing in the union, and so an Id is not required. Maybe that's why TypeObject for unions has a separate field for discriminator that doesn't contain Id. If that's the case, we can use an arbitrary value for the discriminator Id when serializing.

There's a related spec issue to this: https://issues.omg.org/issues/DDSXTY14-59.

Reply:
It looks like in the generated code this is covered by:
```
if (!strm.write_parameter_id(0, size)) {
```
I suggest changing the `0` for both generated and dynamic cases to a named constant which can have comment describing if/how the spec is lacking.